### PR TITLE
Reduce candidate cache overhead

### PR DESCRIPTION
**Pull request by mourisl**

### CODE REVIEW

File: src/chromap.cc [hunk 1]
```diff
@@ -1146,14 +1146,16 @@ void Chromap<MappingRecord>::MapPairedEndReads() {
                 negative_candidates2_buffer.clear();
                 uint32_t repetitive_seed_length1 = 0;
                 uint32_t repetitive_seed_length2 = 0;
+
```
Comment (haowenz) [inline src/chromap.cc:1149]: Can we remove these two extra lines?

File: src/chromap.cc [hunk 2]
```diff
@@ -1327,22 +1329,35 @@ void Chromap<MappingRecord>::MapPairedEndReads() {
...
-          for (uint32_t pair_index = 0; pair_index < num_loaded_pairs / 2; ++pair_index) {
-            if (num_reads_ >= 2 * 5000000 && pair_index >= num_loaded_pairs / num_threads_) {
+          for (uint32_t pair_index = 0; pair_index < num_loaded_pairs; ++pair_index) {
+            if (num_reads_ > 5000000 && pair_index >= num_loaded_pairs / num_threads_) {
               break;
             }
             mm_to_candidates_cache.Update(mm_history1[pair_index].minimizers, mm_history1[pair_index].positive_candidates, mm_history1[pair_index].negative_candidates, mm_history1[pair_index].repetitive_seed_length);
             mm_to_candidates_cache.Update(mm_history2[pair_index].minimizers, mm_history2[pair_index].positive_candidates, mm_history2[pair_index].negative_candidates, mm_history2[pair_index].repetitive_seed_length);
-            if (mm_history1[pair_index].positive_candidates.size() < mm_history1[pair_index].positive_candidates.capacity() / 2) {
+            /*if (mm_history1[pair_index].positive_candidates.size() < mm_history1[pair_index].positive_candidates.capacity() / 50) {
```
Comment (haowenz) [inline src/chromap.cc:1342]: Why these are commented? Can we remove them?

File: src/chromap.cc [hunk 3]
```diff
@@ -1327,22 +1329,35 @@ void Chromap<MappingRecord>::MapPairedEndReads() {
...
+              std::vector<Candidate>().swap(mm_history1[pair_index].negative_candidates);
+            }
+            if (mm_history2[pair_index].positive_candidates.size() < mm_history2[pair_index].positive_candidates.capacity() / 50) {
+              std::vector<Candidate>().swap(mm_history2[pair_index].positive_candidates);
+            }
+            if (mm_history2[pair_index].negative_candidates.size() < mm_history2[pair_index].negative_candidates.capacity() / 50) {
+              std::vector<Candidate>().swap(mm_history2[pair_index].negative_candidates);
+            }*/
+
+						if (mm_history1[pair_index].positive_candidates.size() > 50) {
```
Comment (haowenz) [inline src/chromap.cc:1355]: Can we fix the indentation of this if?

File: src/chromap.cc [hunk 4]
```diff
@@ -1389,6 +1404,7 @@ void Chromap<MappingRecord>::MapPairedEndReads() {
   std::cerr << "Mapped all reads in " << Chromap<>::GetRealTime() - real_start_mapping_time << "s.\n";
   delete[] mm_history1;
   delete[] mm_history2;
+	//mm_to_candidates_cache.PrintStats();
```
Comment (haowenz) [inline src/chromap.cc:1411]: Can we remove this comment?

File: src/chromap.cc [hunk 5]
```diff
@@ -1327,22 +1329,35 @@ void Chromap<MappingRecord>::MapPairedEndReads() {
...
+            }
+            if (mm_history2[pair_index].negative_candidates.size() < mm_history2[pair_index].negative_candidates.capacity() / 50) {
+              std::vector<Candidate>().swap(mm_history2[pair_index].negative_candidates);
+            }*/
+
+						if (mm_history1[pair_index].positive_candidates.size() > 50) {
               std::vector<Candidate>().swap(mm_history1[pair_index].positive_candidates);
             }
-            if (mm_history1[pair_index].negative_candidates.size() < mm_history1[pair_index].negative_candidates.capacity() / 2) {
+            if (mm_history1[pair_index].negative_candidates.size() > 50) {
```
Comment (haowenz) [inline src/chromap.cc:1358]: Can you clarify what these swaps do and why use 50 as a threshold?

**Review by haowenz — CHANGES_REQUESTED** [review]

File: src/chromap.cc (hunk 4 above)
Comment (mourisl) [reply src/chromap.cc]: The code mentioned above is all cleaned.

File: src/chromap.cc (hunk 5 above)
Comment (mourisl) [reply src/chromap.cc:1344]: This is used to release the memory of the candidate list. I think 50 is roughly a good number. It does not release too often and also avoids the case with too many candidates which wastes the memory.

**Review by haowenz — APPROVED** [review]

**Pull request merged**